// angular-http/src/app/amirada/components/mask/mask.component.ts
import {Component, OnInit, Input, ViewChildren, QueryList, AfterViewInit, DoCheck} from '@angular/core';
import { UDFModel } from '../../models/udf.model';


@Component({
  selector: 'pf-mask-ctrl',
  template: `
    <div class="mask-ctrl">
      <div class="mask-ctrl-btn">
        <a (click)="active = !active">{{active ? '&#9660;' : '&#9658;'}}
          <span [innerHTML]="model.desc | safeValue"></span>
        </a>
      </div>
      <div *ngIf="active">
        <ng-container *ngFor="let item of model.children">
          <div class="mask-io">
            <div class="mask-io-elm" *ngFor="let value of model.query('value')">
              <input [id]="value.uid" [type]="value.uiType" [value]="value.value"
                     [checked]="value.active"
                     [disabled]="value.disabled"
                     (change)="value.active = !value.active">
              <label for="value.uid" [style.opacity]="value.disabled ? '.3' : '1'">
                <span [innerHTML]="value.desc | safeValue"></span> ({{value.numHits}})
              </label>
            </div>
          </div>
        </ng-container>
      </div>
    </div>
  `,
  styles: [
    `.mask-ctrl {border: 0px solid #FF0000; margin: 2px; padding: 2px; background-color: #666666; color: #FFF;}  
    .mask-ctrl-btn {height: 32px; display: flex; flex-direction: column; justify-content: center}
    .mask-ctrl-btn > a {cursor: pointer; width:100%; display: block;}
    .mask-io { border: 0px solid #ff8720; margin: 2px; padding: 0px; background-color: #FFF;  color: #000; }
    .mask-io-elm { border-bottom: 1px solid #666; margin: 0px; padding: 2px; }
    `
  ]
})
export class MaskCtrlComponent implements OnInit, DoCheck  {
  @Input() model: UDFModel;
  active = true;
  constructor() { }
  ngOnInit() {
    // console.log('this.model', this.model);
  }
  ngDoCheck() {
    // console.log('MaskCtrlComponent changes', this.model.constructor.name, this.model.toString());
  }
}


@Component({
  selector: 'pf-mask',
  template: `
    <h3>Mask Component</h3>
    <div *ngIf="model">
      {{model.desc}}
      <div class="mask-controls">
        <ng-container *ngFor="let item of model.query('compound compound')">
          <pf-mask-ctrl *ngIf="!item.hidden" [model]="item"></pf-mask-ctrl>
        </ng-container>
      </div>
    </div>
  `,
  styleUrls: ['./mask.component.css']
})
export class MaskComponent implements OnInit, DoCheck {

  @Input() model: UDFModel;
  @ViewChildren(MaskCtrlComponent) maskControles: QueryList<MaskCtrlComponent>;
  constructor() { }

  ngOnInit() {
    // this.model.assignedComponent = 'compound';
  }

  ngDoCheck() {
    // console.log('MaskComponent changes', this.model.constructor.name, this.model.name);
  }

}
