// angular-http/src/app/amirada/components/paramsmenu/paramsmenu.component.ts
import {Component, OnInit, OnChanges, AfterViewChecked, AfterViewInit, Output, EventEmitter} from '@angular/core';
import { PfDataService, OptionList, IOptionListItem } from '../../data.service';
import {Observable} from 'rxjs/Observable';

export interface IParams {
  lcc?: IOptionListItem;
  lcd?: IOptionListItem;
  frq?: IOptionListItem;
}

@Component({
  selector: 'pf-paramsmenu',
  templateUrl: './paramsmenu.component.html',
  styleUrls: ['./paramsmenu.component.css']
})
export class PfLangmenuComponent implements OnInit, OnChanges, AfterViewChecked, AfterViewInit  {
  public countries: OptionList;
  public languages: OptionList;
  public frequencies: OptionList;
  public params: IParams = {};

  @Output() paramsChange = new EventEmitter();

  constructor(private pfinderService: PfDataService ) {
    // Nada
  }

  ngOnInit() {
    // this.countries = this.pfinderService.getCountries();
    this.pfinderService.getCountries()
      .subscribe(res => this.countries = res);

    this.pfinderService.getLanguages()
      .subscribe(res => this.languages = res);

    this.pfinderService.getHz()
      .subscribe(res => this.frequencies = res);

  }

  setParam(value: IOptionListItem) {
    this.params[value.type] = value;
  }

  sendParams() {
    const params = {};
    let key = '';
    for (key in this.params) {
       params[this.params[key].type] = this.params[key].value;
    }
    if (key /* there es at least one param */ ) {
      this.paramsChange.emit(params);
      // this.pfinderService.login(params);
    }
  }

  ngOnChanges(changes) {
    // console.log('LangmenuComponent#ngOnChanges changes=', changes);
  }

  ngAfterViewInit() {
    // console.log('LangmenuComponent#ngAfterViewInit');
  }
  ngAfterViewChecked() {
    // console.log('LangmenuComponent#ngAfterViewChecked');
  }

}
