// angular-http/src/app/amirada/components/list/list.component.ts
import {Component, Input, OnInit, DoCheck} from '@angular/core';
import { UDFModel } from '../../models/udf.model';

@Component({
  selector: 'pf-list',
  template: `
    <h3>List Component</h3>
    <div *ngIf="model && model.query('compound compound').length">
      {{model.desc}}
      <select [(ngModel)]="sortBy" (ngModelChange)="setSort()">
        <option *ngFor="let column of model.query('compound compound')[0].header" 
                type="radio" id="{{column.name}}" value="{{column.name}}">
          <span [innerHTML]="column.desc | safeValue"></span>
        </option>
      </select>
      <select [(ngModel)]="sortOrder" (ngModelChange)="setSort()">
        <option type="radio" value="asc">asc</option>
        <option type="radio" value="desc">desc</option>
      </select>
    </div>
    <ng-container *pfDynComp="model"></ng-container>
  `,
  styles: []
})
export class ListComponent implements OnInit, DoCheck {

  @Input() model: UDFModel;
  sortBy: string;
  sortOrder = 'asc';

  constructor() { }

  setSort() {
    if (this.sortBy) {
      this.model.context.sort_by = this.sortBy;
      this.model.context.sort_order = this.sortOrder;
    }
  }

  ngOnInit() {
    this.model.assignedComponent = 'compound';
  }

  ngDoCheck() {
    // console.log('MaskComponent changes', this.model.constructor.name, this.model.name);
  }

}
